common/datasets/sms_wsj/returnn_datasets.py

```diff
@@ -0,0 +1,617 @@
+        :param Callable segment_mapping_fn: function that maps SMS-WSJ seg. name into list of corresp. seg. names in HDF
+        :param Optional[int] pad_label: target label assigned to padded areas
+        :param str hdf_data_key: data key under which the alignment is stored in the hdf, usually "classes" or "data"
+        :param kwargs:
+        """
+        super().__init__(**kwargs)
+
+        self._hdf_dataset = HDFDataset([hdf_file], use_cache_manager=True)
+        self._segment_mapping_fn = segment_mapping_fn
+        self._pad_label = pad_label
+        self._hdf_data_key = hdf_data_key
+
+    def __getitem__(self, seq_idx: int) -> Dict[str, np.array]:
+        d = self._get_seq_by_idx(seq_idx)
+        hdf_seq_tags = self._segment_mapping_fn(str(d["seq_tag"]))
+        assert (
+            len(hdf_seq_tags) == d["target_signals"].shape[1]
+        ), f"got {len(hdf_seq_tags)} segment names, but there are {d['target_signals'].shape[1]} target signals"
+        hdf_classes = [
+            self._hdf_dataset.get_data_by_seq_tag(hdf_seq_tag, self._hdf_data_key)
+            for hdf_seq_tag in hdf_seq_tags
+        ]
+        padded_len = max(hdf_classes_.shape[0] for hdf_classes_ in hdf_classes)
+        for speaker_idx, hdf_classes_speaker in enumerate(hdf_classes):
+            total_pad_frames = padded_len - hdf_classes_speaker.shape[0]
+            if total_pad_frames == 0:
+                continue
+            pad_start = int(round(d["offset"][speaker_idx] / d["seq_len"] * padded_len))
+            pad_start = min(pad_start, total_pad_frames)
+            pad_end = total_pad_frames - pad_start
+            if pad_start or pad_end:
+                assert self._pad_label is not None, "Label for padding is needed"
+            hdf_classes[speaker_idx] = np.concatenate(
+                [
+                    self._pad_label * np.ones(pad_start),
+                    hdf_classes[speaker_idx],
+                    self._pad_label * np.ones(pad_end),
+                ]
+            )
+        d["target_classes"] = np.stack(hdf_classes).T
+        d["target_classes_len"] = np.array(padded_len)
+        return d
+
+    def get_seq_length_for_keys(self, seq_idx: int) -> NumbersDict:
+        """
+        Returns sequence length for all data/target keys.
+        """
+        d = super().get_seq_length_for_keys(seq_idx)
+        data = self[seq_idx]
+        d["target_classes"] = int(data["target_classes_len"])
+        return NumbersDict(d)
+
+
+class SmsWsjWrapper(MapDatasetWrapper):
+    """
+    Base class for datasets that can be used in RETURNN config.
+    """
+
+    def __init__(self, sms_wsj_base, **kwargs):
+        """
+        :param Optional[SmsWsjBase] sms_wsj_base: SMS-WSJ base class to allow inherited classes to modify this
+        """
+        if "seq_ordering" not in kwargs:
+            print("Warning: no shuffling is enabled by default", file=returnn_log.v2)
```

Comment:
```suggestion
        if "seq_ordering" not in kwargs:
            print("Warning: no shuffling is enabled by default", file=returnn_log.v2)
        else:
            self._seq_ordering = kwargs["seq_ordering"]
```

Reply:
Ok, I'm sorry, what I said was not 100% right. I recreated the problem. The issue is not `update_buffer`. 
In my case, `self.seq_ordering` was changed to `"sorted_reverse"`. The `init_seq_order()` function in `MapDatasetWrapper` calls `self.get_seq_order_for_epoch`. If `self.seq_ordering` is `"sorted_reverse"` `get_seq_len()` is called and it tries to get the sequence length for sequences which are not in the buffer. Therefore, `self.seq_ordering` has to be `"default"`, so not all sequence lengths are needed. Maybe there is a better and more elegant solution to this problem, but this was the easiest one I could think of

Reply:
See other thread https://github.com/rwth-i6/i6_experiments/pull/116#discussion_r1103830555